File: library/BackEnd/routes/BookRoute.py

"""
Book Route 
Handles routing for book related operations
"""

from fastapi import APIRouter, Request
from models.BookModel import Book
from models.BookModel import load_books,write_books

router=APIRouter()


"""
Fetches books details 

Returns:
    dict:Book data
"""
@router.get("/books")
def get_Books():
    return load_books()

@router.post("/books")
def add_book(book: Book):
    data = load_books()

    books = data["books"]

    newId = books[-1]["id"] + 1 if books else 1
    newBook = book.model_dump()
    newBook["id"] = newId

    books.append(newBook)
    data["books"] = books

    write_books(data)

    return {"message": "Book added ucessfully"}



@router.delete("/books")
async def delete_Book(request: Request):
    data = await request.json()
    bookid = data.get("id")

    booksData = load_books()
    books = booksData["books"]

    for i, book in enumerate(books):
        if book["id"] == bookid:
            deleteBook = books.pop(i)
            write_books(booksData)
            return {"message": "Book delted suicessfully", "deltedbook": deleteBook}

File: library/BackEnd/routes/BookRoute.js

const express = require("express");
const router=express.Router();
const fs = require("fs");

const {readFile,writeFile}=require("../models/BookModel")


router.get("/books", async (req, res) => {
    res.json(await readFile('data/BookData.json'));
});


router.post('/books', async (req, res) => {
    try {
        const newBook = await req.body;
        console.log(newBook);

        let BookData = await readFile("data/BookData.json");
        let BookList = BookData.books;

        const newId = BookList.length > 0 ? BookList[BookList.length - 1].id + 1 : 1;
        newBook.id = newId;
        console.log(newBook);
        BookData.books.push(newBook);

        await writeFile("data/BookData.json", BookData);
        res.json({ message: "Book Added Sucessfully", newBook });
    } catch (error) {
        res.status(error.status || 500).json({ message: error.message });
    }
});

router.delete("/books", async (req, res) => {
    try {
        const { id } = req.body;
        if (!id) return res.status(400).json({ message: "Book ID is required" });

        const BookData = await readFile("data/BookData.json");
        const AllocationData = await readFile("data/Allocation.json")

        const BookList = BookData.books.filter(book => book.id != id);
        const AllocationList = AllocationData.allocation.filter(book => book.bookId != id);

        if (BookList.length === BookData.books.length) {
            return res.status(404).json({ message: "Book not found" });
        }

        BookData.books = BookList;
        AllocationData.allocation = AllocationList;

        await writeFile("data/BookData.json", BookData);
        await writeFile("data/Allocation.json", AllocationData);
        res.json({ message: "Book Delted Sucessfully ", id });

    } catch (error) {
        res.status(error.status || 500).json({ message: error.message });
    }
});


module.exports=router;